### PR TITLE
Docker SSI tests

### PR DESCRIPTION
**Pull request by robertomonteromiguel**
## Motivation

Related with this task: https://datadoghq.atlassian.net/browse/APMON-1318
We want to bring guardrail tests and service naming tests into system-test.

We create a matrix of weblogs using the new utility class: utils/docker_ssi/docker_ssi_matrix_builder.py
One weblog can be deployed:
- With different language runtimes (ie java 8, java 11...)
- With different architecture (ie arm64 or amd64)
- With different base images (ie ubuntu 22, ubuntu 18)

The weblog are going to be built at runtime when we configure the scenario
**TODOs (Before merge this PR)**
- Push the images only if the tests passed (check revision)
- I broke the docker mode scenario. Why?
- Execute all scenarios, to check if this PR causes problem in other scenarios (parametric? k8s lib injection validation?)
- Check that compute scenarios works well

**TODOs (After merge this PR)**
- Report results to FPD 
- Create a gitlab job, that it can be executed manually and forces the recreating of all docker ssi images.
- Add docs
- Bring the rest of the test (for node, python....)

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

File: utils/_context/_scenarios/docker_ssi.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+    def close_targets(self):
+        self._test_agent_polling.stop()
+        for container in reversed(self._required_containers):
+            try:
+                container.remove()
+                logger.info(f"Removing container {container}")
+            except:
+                logger.exception(f"Failed to remove container {container}")
+        # TODO push images only if all tests pass
+        # self.ssi_image_builder.push_base_image()
```
Comment (robertomonteromiguel) [inline utils/_context/_scenarios/docker_ssi.py:152]: Enable this line before merge only for tests that passed

File: utils/scripts/compute_impacted_scenario.py [hunk 2]
```diff
@@ -174,6 +176,11 @@ def main():
             if file in scenarios_by_files:
                 scenarios.update(scenarios_by_files[file])
 
+    # print("scenarios=" + ",".join(scenarios))
+    # print("scenarios_groups=" + ",".join(scenarios_groups))
+    # TODO RMM REMOVE THIS
+    scenarios = set(["DOCKER_SSI"])
+    scenarios_groups = set([ScenarioGroup.DOCKER_SSI.value])
```
Comment (robertomonteromiguel) [inline utils/scripts/compute_impacted_scenario.py:183]: Fix this before merge!!!

File: utils/_context/containers.py [hunk 3]
```diff
@@ -1072,3 +1074,34 @@ def set_environment_for_library(self, library):
...
+            host_log_folder=host_log_folder,
+            ports={"18080": ("127.0.0.1", 18080), "8080": ("127.0.0.1", 8080)},
+            healthcheck={"test": "sh /healthcheck.sh", "retries": 60,},
+            allow_old_container=False,
+            environment={"DD_DEBUG": "true", "DD_TRACE_SAMPLE_RATE": 1, "DD_TELEMETRY_METRICS_INTERVAL_SECONDS": "0.5"},
+            volumes={f"./{host_log_folder}/interfaces/test_agent_socket": {"bind": "/var/run/datadog/", "mode": "rw",}},
+        )
+
+    def get_env(self, env_var):
+        """Get env variables from the container (only works for started containers)"""
```
[the remark this answers is not in the document]
Comment (robertomonteromiguel) [reply utils/_context/containers.py]: done!

File: utils/docker_ssi/test_agent_polling.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import pathlib
```
[the remark this answers is not in the document]
Comment (robertomonteromiguel) [reply utils/docker_ssi/test_agent_polling.py]: done!

File: tests/docker_ssi/test_docker_ssi.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        self.r = TestDockerSSIFeatures._r
+
+    def setup_install(self):
+        self._setup_all()
+
+    @features.ssi_guardrails
+    @bug(
+        condition="centos-7" in context.scenario.weblog_variant and context.scenario.library.library == "java",
+        reason="There is a issue building the image on centos 7",
+    )
```
[the remark this answers is not in the document]
Comment (robertomonteromiguel) [reply tests/docker_ssi/test_docker_ssi.py]: done!

File: utils/_context/_scenarios/docker_ssi.py [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+
+    def close_targets(self):
+        self._test_agent_polling.stop()
+        for container in reversed(self._required_containers):
+            try:
+                container.remove()
+                logger.info(f"Removing container {container}")
+            except:
+                logger.exception(f"Failed to remove container {container}")
+        # TODO push images only if all tests pass
```
[the remark this answers is not in the document]
Comment (robertomonteromiguel) [reply utils/_context/_scenarios/docker_ssi.py:123]: TODO after merge. Jira ticket: https://datadoghq.atlassian.net/browse/APMON-1482

**Review by cbeauchesne — APPROVED** [review]
Approved, but there is warning we'll need to fix : 
image

**Comment by robertomonteromiguel** [comment]
> Approved, but there is warning we'll need to fix : image

I'll solve in a next PR

**Pull request merged**